src/complex_fluids/AdvDiffComplexFluidConvectiveOperator.cpp

```diff
@@ -0,0 +1,464 @@
+      d_Q_var(Q_var),
+      d_u_adv_var(new SideVariable<NDIM, double>("Complex U var")),
+      d_difference_form(difference_form),
+      d_conc_bc_coefs(conc_bc_coefs),
+      d_u_bc_coefs(u_bc_coefs)
+{
+    d_interp_type = input_db->getStringWithDefault("interp_type", d_interp_type);
+    d_sqr_root = input_db->getBoolWithDefault("square_root_evolve", false);
+    d_log_conform = input_db->getBoolWithDefault("log_conform_evolve", false);
+    // Register some scratch variables
+    auto var_db = VariableDatabase<NDIM>::getDatabase();
+    Pointer<VariableContext> context = var_db->getContext(d_object_name + "::CONTEXT");
+    d_Q_convec_idx = var_db->registerVariableAndContext(
+        d_Q_var, var_db->getContext(d_object_name + "::CONVECTIVE"), IntVector<NDIM>(0));
+    Pointer<VariableContext> rhs_ctx = var_db->getContext(d_object_name + "::RHS_CONTEXT");
+    d_R_idx = var_db->registerVariableAndContext(d_Q_var, rhs_ctx);
+    Pointer<VariableContext> new_cxt = var_db->getContext(d_object_name + "::U_ADV_CXT");
+    d_u_scratch_idx = var_db->registerVariableAndContext(d_u_adv_var, new_cxt, IntVector<NDIM>(7));
+    switch (d_difference_form)
+    {
+    case CENTERED:
+        d_convec_oper = new AdvDiffCenteredConvectiveOperator(
+            d_object_name + "::convec_oper", d_Q_var, input_db, ADVECTIVE, conc_bc_coefs);
+        break;
+    case WAVE_PROP:
+        d_convec_oper = new AdvDiffWavePropConvectiveOperator(
+            d_object_name + "::convec_oper", d_Q_var, input_db, ADVECTIVE, conc_bc_coefs);
+        break;
+    case CF_PPM:
+        d_convec_oper = new AdvDiffPPMConvectiveOperator(
+            d_object_name + "::convec_oper", d_Q_var, input_db, ADVECTIVE, conc_bc_coefs);
+        break;
+    case UNKNOWN:
+        TBOX_ERROR("\n\n CONVECTIVE OPERATOR: " + enum_to_string<CFConvectiveOperatorType>(d_difference_form) +
+                   " UNKNOWN\n\n");
+        break;
+    }
+} // Constructor
+
+AdvDiffComplexFluidConvectiveOperator::~AdvDiffComplexFluidConvectiveOperator()
+{
+    deallocateOperatorState();
+    return;
+} // Destructor
+
+void
+AdvDiffComplexFluidConvectiveOperator::applyConvectiveOperator(int Q_idx, int Y_idx)
+{
+    if (!d_rhs)
+    {
+        TBOX_ERROR("AdvDiffComplexFluidConvectiveOperator::applyConvectiveOperator():\n"
+                   << "  Relaxation function must be register prior to call to applyConvectiveOperator\n");
+    }
+    if (!d_is_initialized)
+    {
+        TBOX_ERROR("AdvDiffComplexFluidConvectiveOperator::applyConvectiveOperator():\n"
+                   << "  operator must be initialized prior to call to applyConvectiveOperator\n");
+    }
+    // Set up velocity information:
+    d_convec_oper->setAdvectionVelocity(d_u_idx);
+    d_convec_oper->setSolutionTime(d_solution_time);
+    Pointer<CartesianGridGeometry<NDIM>> grid_geom = d_hierarchy->getGridGeometry();
+    // Copy data to side centered velocity field
```

Comment:
We should probably have a helper function to do this kind of thing in IBTK (e.g. in `PatchMathOps`).